src/handlers/permissions-grant.ts

```diff
@@ -0,0 +1,60 @@
+import type { BaseMessageReply } from '../core/message-reply.js';
+import type { MethodHandler } from '../types/method-handler.js';
+import type { PermissionsGrantMessage } from '../types/permissions-types.js';
+import type { DidResolver, EventLog, MessageStore } from '../index.js';
+
+import { authenticate } from '../core/auth.js';
+import { Message } from '../core/message.js';
+import { messageReplyFromError } from '../core/message-reply.js';
+import { PermissionsGrant } from '../interfaces/permissions-grant.js';
+
+export class PermissionsGrantHandler implements MethodHandler {
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private eventLog: EventLog) { }
+
+  public async handle({
+    tenant,
+    message
+  }: { tenant: string, message: PermissionsGrantMessage }): Promise<BaseMessageReply> {
+    let permissionsGrant: PermissionsGrant;
+    try {
+      permissionsGrant = await PermissionsGrant.parse(message);
+    } catch (e) {
+      return messageReplyFromError(e, 400);
+    }
+
+    try {
+      await authenticate(message.authorization, this.didResolver);
+      await permissionsGrant.authorize();
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // validate PermissionsRequestId if it is included
+    if (message.descriptor.permissionsRequestId !== undefined) {
```

Comment:
For my learning: why is `permissionsRequestId` optional?

Reply:
Removed.